### ISSUE TITLE
Nested async functions

### ISSUE THREAD

**Issue by rb1147x** [basro/hx-jsasync#5]
Hi @basro 

I'm exploring your library and came across a problem of using nested `async` functions. It seems like the macro only generates the first `async`.

Consider the following haxe code:

```
import jsasync.JSAsyncTools.jsawait;
import jsasync.IJSAsync;

class Main implements IJSAsync {
	static function main() {
		do_thing();
	}

	@:jsasync static function do_thing() {
		our_promise().then(@:jsasync function(v) {
			var v2 = jsawait(our_promise());
			trace(v2);
		});
	}

	static function our_promise() {
		return new js.lib.Promise(function(resolve, reject) {
			resolve('promise completed');
		});
	}

}

```

When I initially ran into this problem I didn't even have the second `@:jsasync` identifier on my anonymous function, but am including it here to show that it just gets removed in the JS output:

```
// Generated by Haxe 4.1.5
(function ($global) { "use strict";
var Main = function() { };
Main.main = function() {
	Main.do_thing();
};
Main.do_thing = async function() {
	
        // missing async identifier below

	Main.our_promise().then(function(v) {
		console.log("src/Main.hx:12:",(await Main.our_promise()));
	});
};
Main.our_promise = function() {
	return new Promise(function(resolve,reject) {
		resolve("promise completed");
	});
};
var haxe_iterators_ArrayIterator = function(array) {
	this.current = 0;
	this.array = array;
};
haxe_iterators_ArrayIterator.prototype = {
	hasNext: function() {
		return this.current < this.array.length;
	}
	,next: function() {
		return this.array[this.current++];
	}
};
var jsasync_JSAsyncTools = function() { };
var jsasync_impl_Helper = function() { };
Main.main();
})({});
```

And so we then receive the JS error:

```
console.log("src/Main.hx:12:",(await Main.our_promise()));

SyntaxError: Unexpected identifier
```

I'm still learning and doing my best to understand promises. I believe we can refactor the explicit promise out into another `await` so this works:

```
var v = jsawait(our_promise());
var v2 = jsawait(our_promise2(v));
trace(v2);

/*
our_promise().then(@:jsasync function(v) {
var v2 = jsawait(our_promise());
trace(v2);
});
*/
```

I'm not sure if your library already looks for this and it's a bug or would be new functionality. I assumed incorrectly that any child functions within a parent async function are automatically marked async, but that's apparently not the case.

If it's new, would you please consider adding this :)

Thanks.

**Comment by basro** [comment]
Hello,
`@:jsasync` metadata is only used to mark class methods, to create anonymous async functions you have to use the `JSAsync.jsasync` macro.

Like this:
```haxe
import jsasync.JSAsync.jsasync;

/* ... */

var myAsyncFunc = jsasync(function(val:Int) {
    return val;
});
```

It is expected and normal behaviour that @:jsasync will be ignored inside code expressions.

**Comment by rb1147x** [comment]
Ahhh, I see. I missed that looking through the Usage section, thank you for clarifying.

I received some very odd errors if I did not annotate my function parameter, e.g.:

```
jsasync(function(conn) {
// ...
});
```

however, once I added an annotation e.g. `conn:Connection` it seemed to work just fine.

I do randomly receive uncaught exceptions and stack errors pointing to the library, but everything still seems to compile just fine...
